File: scraper.py
# -*- coding: utf-8 -*-
"""
empik-zapowiedzi - scraper zapowiedzi ksiazkowych Empiku dla panelu zaczytajsie.pl

Dziala tak:
1. Pobiera strony listingu https://www.empik.com/zapowiedzi/ksiazki i zbiera ID produktow.
2. Dla produktow nowych lub bez kompletu danych (details_ok=false) pobiera strone produktu
   i wyciaga: tytul, autora, okladke, EAN, date premiery, opis, kategorie.
   Zrodla: JSON-LD (glowne) + meta og:/product: (fallback) + regexy (ostateczny fallback).
   Budzet: max DETAILS_BUDGET stron produktow na jeden przebieg.
3. Zapisuje do tabeli empik_proposals w Supabase:
   - nowe pozycje: INSERT ze status='new'
   - istniejace bez kompletu: UPDATE szczegolow (NIE rusza statusu - added/dismissed zostaja)

Sekrety (GitHub Actions): SUPABASE_URL, SUPABASE_SERVICE_KEY
"""

import json
import os
import re
import sys
import time
from datetime import date, datetime

import requests

# ============ KONFIGURACJA ============
LISTING_URL = "https://www.empik.com/zapowiedzi/ksiazki"
LISTING_PAGES = 3          # ile stron listingu na przebieg (start=0,60,120)
LISTING_STEP = 60          # przesuniecie paginacji (?start=)
DETAILS_BUDGET = 50        # max stron produktow na przebieg (nowe + backfill)
REQUEST_DELAY = 1.5        # sekundy miedzy zadaniami (nie mecz serwera)
EXCLUDED_CATEGORIES = []   # np. ["Komiks"] - kategorie Empiku do pominiecia (puste = bierz wszystko)

SUPABASE_URL = os.environ.get("SUPABASE_URL", "").rstrip("/")
SUPABASE_KEY = os.environ.get("SUPABASE_SERVICE_KEY", "")

# Przegladarowe naglowki - Empik potrafi odrzucac "gole" klienty HTTP (403).
HEADERS = {
    "User-Agent": ("Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 "
                   "(KHTML, like Gecko) Chrome/126.0.0.0 Safari/537.36"),
    "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
    "Accept-Language": "pl-PL,pl;q=0.9,en;q=0.8",
    "Referer": "https://www.empik.com/",
}

session = requests.Session()
session.headers.update(HEADERS)


def log(*args):
    print("[empik-zapowiedzi]", *args, flush=True)


def fetch(url):
    """Pobierz strone; zwroc HTML albo None. Loguje status - kluczowe przy debugowaniu 403."""
    try:
        r = session.get(url, timeout=30)
        if r.status_code != 200:
            log(f"HTTP {r.status_code} dla {url}")
            return None
        return r.text
    except requests.RequestException as e:
        log(f"Blad sieci dla {url}: {e}")
        return None


# ============ LISTING: zbieranie ID produktow ============

# Link produktu-ksiazki: https://www.empik.com/<slug>,p<ID>,ksiazka-p
RE_PRODUCT = re.compile(r'href="(https://www\.empik\.com/[^"]*?,p(\d+),ksiazka-p)[^"]*"')


def collect_listing_ids():
    """Zwraca dict {empik_id: url} z pierwszych LISTING_PAGES stron listingu."""
    found = {}
    for page in range(LISTING_PAGES):
        url = LISTING_URL if page == 0 else f"{LISTING_URL}?start={page * LISTING_STEP}"
        html = fetch(url)
        if not html:
            break
        before = len(found)
        for m in RE_PRODUCT.finditer(html):
            found.setdefault(m.group(2), m.group(1))
        log(f"listing strona {page + 1}: {len(found) - before} nowych linkow (razem {len(found)})")
        if len(found) == before and page > 0:
            break  # paginacja nie dziala albo koniec - nie brnij dalej
        time.sleep(REQUEST_DELAY)
    return found


# ============ STRONA PRODUKTU: szczegoly ============

RE_JSONLD = re.compile(r'<script[^>]*type="application/ld\+json"[^>]*>(.*?)</script>', re.DOTALL)
RE_OG = lambda prop: re.compile(r'<meta[^>]+property="og:' + prop + r'"[^>]+content="([^"]*)"')
RE_META_PRODUCT = lambda name: re.compile(r'<meta[^>]+property="product:' + name + r'"[^>]+content="([^"]*)"')
RE_DATE_PL = re.compile(r'premier[a-z]*[:\s]*(?:&nbsp;|\s)*(\d{2})\.(\d{2})\.(\d{4})', re.IGNORECASE)
RE_EAN_TXT = re.compile(r'(?:EAN|ISBN)\D{0,20}(\d{13})')
RE_H1 = re.compile(r'<h1[^>]*>(.*?)</h1>', re.DOTALL)
RE_TAGS = re.compile(r'<[^>]+>')


def strip_tags(s):
    return RE_TAGS.sub(' ', s or '').replace('&amp;', '&').replace('&quot;', '"').replace('&#39;', "'").strip()


def parse_jsonld_blocks(html):
    """Wyciagnij wszystkie bloki JSON-LD (lista dictow; obsluga list i @graph)."""
    out = []
    for m in RE_JSONLD.finditer(html):
        raw = m.group(1).strip()
        try:
            data = json.loads(raw)
        except json.JSONDecodeError:
            continue
        items = data if isinstance(data, list) else data.get('@graph', [data]) if isinstance(data, dict) else []
        for it in items:
            if isinstance(it, dict):
                out.append(it)
    return out


def pl_date_to_iso(d, mth, y):
    try:
        return date(int(y), int(mth), int(d)).isoformat()
    except ValueError:
        return None


def parse_product(html):
    """Zwraca dict szczegolow. Kazde pole: JSON-LD -> meta -> regex. Brakujace = None."""
    det = {"title": None, "author": None, "cover_url": None, "ean": None,
           "release_date": None, "description": None, "category": None}

    # --- JSON-LD (typy Product / Book) ---
    for block in parse_jsonld_blocks(html):
        t = block.get("@type", "")
        types = t if isinstance(t, list) else [t]
        if not any(x in ("Product", "Book") for x in types):
            continue
        det["title"] = det["title"] or strip_tags(block.get("name") or "") or None
        # autor: string, dict albo lista dictow
        a = block.get("author")
        if a and not det["author"]:
            if isinstance(a, list):
                det["author"] = ", ".join(strip_tags(x.get("name", "") if isinstance(x, dict) else str(x)) for x in a) or None
            elif isinstance(a, dict):
                det["author"] = strip_tags(a.get("name", "")) or None
            else:
                det["author"] = strip_tags(str(a)) or None
        img = block.get("image")
        if img and not det["cover_url"]:
            det["cover_url"] = img[0] if isinstance(img, list) else (img.get("url") if isinstance(img, dict) else img)
        det["ean"] = det["ean"] or block.get("gtin13") or block.get("isbn") or None
        rd = block.get("releaseDate") or block.get("datePublished")
        if rd and not det["release_date"]:
            m = re.match(r'(\d{4})-(\d{2})-(\d{2})', str(rd))
            if m:
                det["release_date"] = m.group(0)
        det["description"] = det["description"] or strip_tags(block.get("description") or "") or None
        cat = block.get("category")
        if cat and not det["category"]:
            det["category"] = strip_tags(cat if isinstance(cat, str) else str(cat))

    # --- Fallbacki: meta og:/product: ---
    if not det["cover_url"]:
        m = RE_OG("image").search(html)
        if m:
            det["cover_url"] = m.group(1)
    if not det["title"]:
        m = RE_OG("title").search(html)
        if m:
            det["title"] = strip_tags(m.group(1)) or None
    if not det["category"]:
        m = RE_META_PRODUCT("category2").search(html)
        if m:
            det["category"] = strip_tags(m.group(1)) or None

    # --- Fallbacki: regexy w tresci ---
    if not det["title"]:
        m = RE_H1.search(html)
        if m:
            det["title"] = strip_tags(m.group(1)) or None
    if not det["release_date"]:
        m = RE_DATE_PL.search(html)
        if m:
            det["release_date"] = pl_date_to_iso(m.group(1), m.group(2), m.group(3))
    if not det["ean"]:
        m = RE_EAN_TXT.search(html)
        if m:
            det["ean"] = m.group(1)

    if det["description"]:
        det["description"] = det["description"][:4000]
    return det


def details_complete(det):
    """Komplet = jest tytul, data premiery i opis (reszta mile widziana)."""
    return bool(det.get("title") and det.get("release_date") and det.get("description"))


# ============ SUPABASE ============

def sb_headers(extra=None):
    h = {"apikey": SUPABASE_KEY, "Authorization": f"Bearer {SUPABASE_KEY}",
         "Content-Type": "application/json"}
    if extra:
        h.update(extra)
    return h


def sb_get_existing():
    """Mapa {empik_id: {'status':..., 'details_ok':...}} istniejacych wierszy."""
    url = f"{SUPABASE_URL}/rest/v1/empik_proposals?select=empik_id,status,details_ok&limit=10000"
    r = requests.get(url, headers=sb_headers(), timeout=30)
    r.raise_for_status()
    return {row["empik_id"]: row for row in r.json()}


def sb_insert(rows):
    if not rows:
        return
    url = f"{SUPABASE_URL}/rest/v1/empik_proposals"
    r = requests.post(url, headers=sb_headers({"Prefer": "return=minimal"}),
                      data=json.dumps(rows), timeout=30)
    if r.status_code not in (200, 201):
        log(f"INSERT blad {r.status_code}: {r.text[:300]}")
    else:
        log(f"INSERT: {len(rows)} nowych propozycji")


def sb_update_details(empik_id, det, ok):
    """Aktualizuje TYLKO szczegoly - status pozostaje nietkniety."""
    payload = {k: v for k, v in det.items() if v is not None}
    payload["details_ok"] = ok
    payload["updated_at"] = datetime.utcnow().isoformat()
    url = f"{SUPABASE_URL}/rest/v1/empik_proposals?empik_id=eq.{empik_id}"
    r = requests.patch(url, headers=sb_headers({"Prefer": "return=minimal"}),
                       data=json.dumps(payload), timeout=30)
    if r.status_code not in (200, 204):
        log(f"UPDATE {empik_id} blad {r.status_code}: {r.text[:300]}")


# ============ GLOWNY PRZEBIEG ============

def main():
    if not SUPABASE_URL or not SUPABASE_KEY:
        log("BRAK SEKRETOW: ustaw SUPABASE_URL i SUPABASE_SERVICE_KEY")
        sys.exit(1)

    today = date.today().isoformat()

    listing = collect_listing_ids()
    if not listing:
        log("Listing pusty - prawdopodobnie blokada (sprawdz logi HTTP powyzej). Koncze.")
        sys.exit(1)

    existing = sb_get_existing()
    log(f"W bazie: {len(existing)} propozycji; na listingu: {len(listing)}")

    # Kolejka stron produktow: najpierw zupelnie nowe, potem backfill (details_ok=false)
    new_ids = [pid for pid in listing if pid not in existing]
    backfill_ids = [pid for pid in listing
                    if pid in existing and not existing[pid].get("details_ok")
                    and existing[pid].get("status") == "new"]
    queue = (new_ids + backfill_ids)[:DETAILS_BUDGET]
    log(f"Nowych: {len(new_ids)}, do uzupelnienia: {len(backfill_ids)}, w tym przebiegu: {len(queue)}")

    inserts = []
    for pid in queue:
        html = fetch(listing[pid])
        time.sleep(REQUEST_DELAY)
        if not html:
            continue
        det = parse_product(html)
        ok = details_complete(det)

        if det.get("category") and det["category"] in EXCLUDED_CATEGORIES:
            log(f"p{pid}: kategoria wykluczona ({det['category']}) - pomijam")
            continue
        # Zapowiedz z data w przeszlosci nie jest zapowiedzia - pomijamy nowe takie wpisy
        if det.get("release_date") and det["release_date"] < today and pid not in existing:
            log(f"p{pid}: premiera w przeszlosci ({det['release_date']}) - pomijam")
            continue

        if pid not in existing:
            inserts.append({
                "empik_id": pid, "status": "new", "details_ok": ok,
                "title": det["title"] or f"(bez tytulu p{pid})",
                "author": det["author"], "cover_url": det["cover_url"], "ean": det["ean"],
                "release_date": det["release_date"], "description": det["description"],
                "category": det["category"],
            })
            log(f"p{pid}: NOWA - {det['title']!r} ({det['release_date']}) komplet={ok}")
        else:
            sb_update_details(pid, det, ok)
            log(f"p{pid}: uzupelniono szczegoly, komplet={ok}")

    sb_insert(inserts)
    log("Gotowe.")


if __name__ == "__main__":
    main()

File: scraper.js
// Empik Zapowiedzi → Supabase (zaczytajsie.pl)
// Pobiera listing zapowiedzi Empiku, doczytuje ze stron produktowych komplet
// danych (autor, okładka, EAN, data premiery, opis) i upsertuje do tabeli
// empik_proposals w Supabase. Panel admina czyta z tej tabeli.
//
// Wymagane sekrety (env): SUPABASE_URL, SUPABASE_SERVICE_KEY

import * as cheerio from "cheerio";

const LISTING_URL =
  "https://www.empik.com/zapowiedzi?searchCategory=31&hideUnavailable=true&sort=scoreDesc&availabilitySeparable=przedsprzedaz&qtype=facetForm";

const MAX_PAGES = 5;
const RESULTS_PER_PAGE = 50;
const DETAILS_BUDGET = 50; // max stron produktowych na jeden run

const SUPABASE_URL = (process.env.SUPABASE_URL || "").replace(/\/$/, "");
const SUPABASE_KEY = process.env.SUPABASE_SERVICE_KEY || "";

const HEADERS = {
  "User-Agent":
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/126.0.0.0 Safari/537.36",
  "Accept":
    "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8",
  "Accept-Language": "pl-PL,pl;q=0.9,en;q=0.8",
  "Referer": "https://www.empik.com/",
};

const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

// ---------- Supabase (REST, bez zależności) ----------

async function sb(path, options = {}) {
  const res = await fetch(`${SUPABASE_URL}/rest/v1/${path}`, {
    ...options,
    headers: {
      apikey: SUPABASE_KEY,
      Authorization: `Bearer ${SUPABASE_KEY}`,
      "Content-Type": "application/json",
      ...(options.headers || {}),
    },
  });
  if (!res.ok) {
    const body = await res.text();
    throw new Error(`Supabase ${res.status}: ${body.slice(0, 300)}`);
  }
  const text = await res.text();
  return text ? JSON.parse(text) : null;
}

async function getExisting() {
  // Pobieramy id + details_ok wszystkich rekordów (paczkami po 1000)
  const map = new Map();
  let from = 0;
  while (true) {
    const rows = await sb(
      `empik_proposals?select=id,details_ok&order=id&limit=1000&offset=${from}`
    );
    for (const r of rows) map.set(r.id, r.details_ok);
    if (rows.length < 1000) break;
    from += 1000;
  }
  return map;
}

async function upsertRows(rows) {
  if (rows.length === 0) return;
  await sb(`empik_proposals?on_conflict=id`, {
    method: "POST",
    headers: { Prefer: "resolution=merge-duplicates,return=minimal" },
    body: JSON.stringify(rows),
  });
}

// ---------- Listing ----------

async function fetchListingPage(startPos) {
  const url = `${LISTING_URL}&resultsPP=${RESULTS_PER_PAGE}&start=${startPos}`;
  const res = await fetch(url, { headers: HEADERS });
  if (!res.ok) throw new Error(`HTTP ${res.status} dla start=${startPos}`);
  return await res.text();
}

function parseListing(html) {
  const $ = cheerio.load(html);
  const products = new Map();
  $("a[href*=',p']").each((_, el) => {
    const href = $(el).attr("href") || "";
    const m = href.match(/,(p\d{6,}),/);
    if (!m) return;
    const id = m[1];
    let title =
      $(el).attr("title") ||
      $(el).find("[class*='title'], strong, span").first().text().trim() ||
      $(el).text().trim();
    title = title.replace(/\s+/g, " ").trim();
    const existing = products.get(id);
    if (!existing || (title && title.length > existing.title.length)) {
      products.set(id, {
        id,
        title: title.slice(0, 300),
        url: `https://www.empik.com${href.split("?")[0]}`,
      });
    }
  });
  return products;
}

// ---------- Strona produktu: komplet danych ----------

function normalizeDate(s) {
  if (!s) return null;
  s = s.trim();
  let m = s.match(/^(\d{4})-(\d{2})-(\d{2})/);
  if (m) return `${m[1]}-${m[2]}-${m[3]}`;
  m = s.match(/^(\d{1,2})[.\-\/](\d{1,2})[.\-\/](\d{4})/);
  if (m) return `${m[3]}-${m[2].padStart(2, "0")}-${m[1].padStart(2, "0")}`;
  return null;
}

function cleanText(s, max) {
  return (s || "").replace(/\s+/g, " ").trim().slice(0, max);
}

// Czyści tytuł z og:title: ucina końcówki typu " - Autor", " - Książka",
// oraz sufiks " | Empik.com" / " | Empik".
function cleanOgTitle(raw) {
  let t = cleanText(raw, 400);
  t = t.replace(/\s*\|\s*Empik(\.com)?\s*$/i, "");
  // "Tytuł - Książka" → "Tytuł"
  t = t.replace(/\s*-\s*Książka\s*$/i, "");
  return cleanText(t, 300);
}

// Empik w og:image podaje URL z parametrem rozmiaru (np. ...,w-360).
// Podnosimy do największego dostępnego wariantu.
function upscaleCover(u) {
  if (!u) return u;
  return u
    .replace(/,(?:w|h)-\d+/gi, "")           // usuń parametry rozmiaru w slug
    .replace(/([?&])(?:w|h|width|height|size)=\d+/gi, "$1") // i w query
    .replace(/[?&]+$/, "");
}

async function fetchProductDetails(p) {
  const out = {
    author: null,
    cover_url: null,
    ean: null,
    release_date: null,
    description: null,
    category: null,
    title: null, // ustalamy ze strony produktu (JSON-LD/og:title), nie z listingu
    details_ok: false,
  };
  const SKIP_CRUMBS = new Set([
    "empik", "empik.com", "strona główna", "książki", "ksiazki",
    "zapowiedzi", "przedsprzedaż", "bestsellery", "nowości", "promocje",
  ]);

  const res = await fetch(p.url, { headers: HEADERS });
  if (!res.ok) throw new Error(`HTTP ${res.status}`);
  const html = await res.text();
  const $ = cheerio.load(html);

  // 1) JSON-LD: Product / Book / BreadcrumbList
  $("script[type='application/ld+json']").each((_, el) => {
    let data;
    try {
      data = JSON.parse($(el).contents().text());
    } catch {
      return;
    }
    const nodes = [].concat(data).flatMap((d) => (d && d["@graph"] ? d["@graph"] : [d]));
    for (const node of nodes) {
      if (!node || !node["@type"]) continue;
      const types = [].concat(node["@type"]);

      if (types.includes("Product") || types.includes("Book")) {
        if (node.name && !out.title) out.title = cleanText(node.name, 300);
        if (node.description) out.description = cleanText(node.description, 3000);
        const img = Array.isArray(node.image) ? node.image[0] : node.image;
        const imgUrl = typeof img === "string" ? img : img?.url || null;
        if (imgUrl && !out.cover_url) out.cover_url = imgUrl;
        // EAN: akceptujemy tylko poprawny 13-cyfrowy kod
        const eanCand = String(node.gtin13 || node.gtin || node.isbn || "").replace(/\D/g, "");
        if (!out.ean && /^\d{13}$/.test(eanCand)) out.ean = eanCand;
        const rd = node.releaseDate || node.datePublished;
        if (rd) out.release_date = normalizeDate(String(rd)) || out.release_date;
        const authors = [].concat(node.author || []).map((a) =>
          typeof a === "string" ? a : a?.name || ""
        );
        const authorStr = authors.filter(Boolean).join(", ");
        if (authorStr && !out.author) out.author = cleanText(authorStr, 200);
      }

      if (types.includes("BreadcrumbList")) {
        const crumbs = (node.itemListElement || [])
          .sort((a, b) => (a.position || 0) - (b.position || 0))
          .map((i) => cleanText(i.name || i.item?.name || "", 60))
          .filter((c) => c && !SKIP_CRUMBS.has(c.toLowerCase()));
        if (crumbs.length > 0) {
          const last = crumbs[crumbs.length - 1];
          out.category =
            crumbs.length >= 2 && last.length > 40 ? crumbs[crumbs.length - 2] : last;
        }
      }
    }
  });

  // 2) Fallbacki i doprecyzowanie z HTML/meta

  // Kategoria: surowa z meta product:category2 (np. "Kryminał, sensacja, thriller").
  // NIE mapujemy tu na gatunki portalu — to robi panel admina.
  if (!out.category) {
    const cat =
      $("meta[property='product:category2']").attr("content") ||
      $("meta[property='product:category']").attr("content");
    if (cat) out.category = cleanText(cat, 80);
  }
  // Fallback kategorii: przedostatni okruszek z breadcrumbów (już policzony wyżej)

  // Tytuł: jeśli JSON-LD nie dał, bierzemy og:title z uciętą końcówką " - Autor"/" - Książka".
  if (!out.title) {
    const ogTitle = cleanOgTitle($("meta[property='og:title']").attr("content"));
    if (ogTitle) out.title = ogTitle;
  }

  // Okładka: og:image z hosta ecsmedia.pl, podniesiona do największego rozmiaru.
  const ogImage = $("meta[property='og:image']").attr("content") || null;
  const coverCandidate = ogImage || out.cover_url;
  out.cover_url = isValidCover(coverCandidate) ? upscaleCover(coverCandidate) : null;

  // Opis: pełny ze strony; og:description tylko jako ostateczność (bywa skrócony).
  if (!out.description) {
    out.description = cleanText(
      $("meta[property='og:description']").attr("content") ||
        $("meta[name='description']").attr("content"),
      3000
    ) || null;
  }

  // EAN: 13 cyfr; szukamy w sekcji szczegółów, potem w całym HTML.
  if (!out.ean) {
    const detail = $("body").text().replace(/\s+/g, " ");
    let m = detail.match(/\bEAN\D{0,15}(\d{13})\b/i) || detail.match(/\bISBN\D{0,15}(\d{13})\b/i);
    if (!m) m = html.match(/\b(97[89]\d{10})\b/);
    if (m) out.ean = m[1];
  }

  // Data premiery: dokładna, format YYYY-MM-DD.
  if (!out.release_date) {
    const flat = html.replace(/\s+/g, " ");
    const m =
      flat.match(/(?:Data premiery|Premiera)\D{0,60}?(\d{4}-\d{2}-\d{2}|\d{1,2}[.\-\/]\d{1,2}[.\-\/]\d{4})/i);
    if (m) out.release_date = normalizeDate(m[1]);
  }

  // Autor: WYŁĄCZNIE z JSON-LD (obsłużone wyżej). Nie pobieramy z breadcrumbów,
  // linków katalogowych ani og:title — tam bywa odwrócona forma "Nazwisko Imię".

  // Jeśli tytuł nadal zawiera doklejoną końcówkę " - Coś" (np. z og:title),
  // ucinamy ją z TYTUŁU, ale NIE traktujemy jako autora.
  if (out.title && out.title.includes(" - ")) {
    const idx = out.title.lastIndexOf(" - ");
    const tail = cleanText(out.title.slice(idx + 3), 200);
    // ucinamy tylko wyraźną końcówkę autorską (krótka, bez cyfr, nie podtytuł/tom)
    if (tail.length <= 60 && !/\d/.test(tail) && !/tom|część|wyd\.|wydanie/i.test(tail)) {
      out.title = cleanText(out.title.slice(0, idx), 300);
    }
  }

  // Kompletność: wpis jest pełny tylko z tytułem, autorem (z JSON-LD),
  // okładką (ecsmedia.pl) i EAN-em. Bez któregokolwiek details_ok=false
  // i kolejne runy spróbują uzupełnić. Lepiej puste niż błędne/odwrócone.
  out.details_ok = Boolean(out.title && out.author && out.cover_url && out.ean);
  return out;
}

// Okładka jest ważna tylko, gdy pochodzi z CDN obrazków Empiku (ecsmedia.pl).
function isValidCover(u) {
  if (!u) return false;
  try {
    return new URL(u).host.includes("ecsmedia.pl");
  } catch {
    return false;
  }
}

// ---------- Main ----------

async function main() {
  if (!SUPABASE_URL || !SUPABASE_KEY) {
    throw new Error("Brak sekretów SUPABASE_URL / SUPABASE_SERVICE_KEY");
  }

  // 1. Listing
  const all = new Map();
  for (let page = 0; page < MAX_PAGES; page++) {
    const start = page * RESULTS_PER_PAGE + 1;
    try {
      const html = await fetchListingPage(start);
      const products = parseListing(html);
      console.log(`Strona start=${start}: ${products.size} produktów`);
      if (products.size === 0) break;
      for (const [id, p] of products) all.set(id, p);
      await sleep(2500 + Math.random() * 2000);
    } catch (e) {
      console.error(`Błąd strony start=${start}: ${e.message}`);
      if (page === 0) throw e;
      break;
    }
  }
  if (all.size === 0) throw new Error("Pusty listing — możliwa blokada/zmiana HTML.");

  // 2. Co już mamy w bazie
  const existing = await getExisting();
  const newOnes = [...all.values()].filter((p) => !existing.has(p.id));
  const needBackfill = [...all.values()].filter(
    (p) => existing.has(p.id) && existing.get(p.id) === false
  );
  console.log(
    `\nListing: ${all.size} | nowych: ${newOnes.length} | do uzupełnienia: ${needBackfill.length}`
  );

  // 3. Nowe bez szczegółów wrzucamy od razu (żeby nic nie zginęło),
  //    szczegóły doczytujemy w ramach budżetu
  await upsertRows(
    newOnes.map((p) => ({ id: p.id, title: p.title, url: p.url }))
  );

  const queue = [...newOnes, ...needBackfill].slice(0, DETAILS_BUDGET);
  let done = 0;
  for (const p of queue) {
    try {
      const d = await fetchProductDetails(p);
      // title jest NOT NULL w bazie — awaryjnie użyj tytułu z listingu
      const safeTitle = d.title || cleanText(p.title, 300);
      await upsertRows([
        {
          id: p.id,
          url: p.url,
          title: safeTitle,
          author: d.author,
          cover_url: d.cover_url,
          ean: d.ean,
          release_date: d.release_date,
          description: d.description,
          category: d.category,
          details_ok: d.details_ok,
          updated_at: new Date().toISOString(),
        },
      ]);
      done++;
      console.log(
        `  ✓ ${safeTitle}${d.author ? ` — ${d.author}` : ""}${d.release_date ? ` (${d.release_date})` : ""}${d.details_ok ? "" : " [niepełne]"}`
      );
    } catch (e) {
      console.error(`  ✗ ${p.title}: ${e.message}`);
    }
    await sleep(1800 + Math.random() * 1200);
  }

  console.log(`\nSzczegóły pobrane: ${done}/${queue.length}`);
  if (queue.length > DETAILS_BUDGET) {
    console.log("Reszta zostanie uzupełniona w kolejnych runach.");
  }
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});
